### DISCUSSION TITLE
Help with modeling a volumetric cable-driven soft robot with Sparsegridtopology

### DISCUSSION THREAD

**Discussion by AndyRay1998** [SofaDefrost/SoftRobots#203] [Help - Using]
Hi,

I am trying to model the bending of a customized cable-driven robot. Since there are too many meshes and the calculation is extremely slow, I used SparseGridTopology to ease the computation. The relevant code are below:
```
cochleaNode.addObject('SparseGridTopology', name='loader', n=[5, 5, 5], fileTopology='mesh/wangshuai.obj')
cochleaNode.addObject('MeshTopology',src = '@loader')
cochleaNode.addObject('MechanicalObject', name='dofs', template='Vec3f', showIndices='false', showIndicesScale='4e-5', rx='0',printLog="0",
                              translation=translation, rotation=rotation)
cochleaNode.addObject('EulerImplicitSolver',rayleighMass=0.1, rayleighStiffness=0.1)
cochleaNode.addObject('CGLinearSolver',iterations="500", tolerance="1e-7" ,threshold="1e-7")
```
and the cable constaints codes, following Finger.py tutorial:
```
nbCables = 3
for i in range(0, nbCables):

        position = [[0., 0., 0.]]*200
        for k in range(0, 200):
                v = Vec3(direction[0]+pullPoint[i][0], direction[1]*0.3*k, direction[2]+pullPoint[i][2])
                position[k] = v


        cableL = cochleaNode.addChild('cableL'+str(i+1))
        cableL.addObject('MechanicalObject', name='dofs',
                                position=pullPoint[i]+[pos.toList() for pos in position])
        cableL.addObject('CableConstraint', template='Vec3', name="cableConstraint"+str(i+1),
                                hasPullPoint='0',
                                indices=list(range(0, 201)),
                                maxPositiveDisp='70',
                                maxDispVariation='1',
                                minForce=0)
        # cableL.addObject("FixedConstraint", indices="-1")
        cableL.addObject('BarycentricMapping', name='mapping',  mapForces=False, mapMasses=False)

```
Nonetheless, I found that the mapping of the cable is calculated towards the center of mass (CoM) of the "first sparse grid unit only", as shown below. When I try to drag the cable, all points of the cable converge towards that one location. Ideally, I wish the cable force is applied to all the FEM units along side the cable constraint routes evenly.
![image](https://user-images.githubusercontent.com/35843385/223073670-71850d50-2ba1-4c32-955d-3054163254b1.png)

Hopefully you can give me some advice on this. Thank you very much! 

All files are shared below.

https://emckclac-my.sharepoint.com/:f:/g/personal/k21025780_kcl_ac_uk/EgPHJV72uFNCpsGMkDISab0BDzOn3MkqteIcv5Io33bVww?e=30z460

**Comment by EulalieCoevoet** [comment]
Hi, 

From the image I can see that you have defined passing points for the cable that are not inside the hexas. 
Try to remove them, it can only work well if each passing point is inside an hexa.
